# **VESPA for PLMs Evaluation**

The aim of this script is to evaluate the generated sequences from PLMs using VESPA.

  

1\. Requirements:
-----------------

*   Python 3.8 or higher
*   Python Libraries: Biopython, Scipy, Seaborn, Matplotlib, Numpy, Pandas, Argsparse
*   VESPA

*To install VESPA and Python Requirements use vespa.yml to create a conda environment:*

```plain
$ conda env create -n vespa --file vespa.yml
$ conda activate vespa 
```
*   Parallel

*To install parallel run this command:*
```
$ sudo apt-get install parallel
```
*   MMSeq2 

*To install MMSeq2:*

```plain
$ git clone https://github.com/soedinglab/MMseqs2.git
$ cd MMseqs2
$ mkdir build
$ cd build
$ cmake -DCMAKE_BUILD_TYPE=RELEASE -DCMAKE_INSTALL_PREFIX=. ..
$ make
$ make install
```
Then add the path of MMSeq2 into the last line the .bashrc file 
```
export PATH="/<Path to MMseq2 directory on your machine>/MMseqs2/build/bin:$PATH"
```
*   ClustalO

*To install ClustalO:*
```
$ sudo apt-get install -y libargtable2-dev
$ wget http://www.clustal.org/omega/clustal-omega-1.2.4.tar.gz
$ tar xvzf clustal-omega-1.2.4.tar.gz
$ cd clustal-omega-1.2.4
$ ./configure CFLAGS="-I/usr/local/include" LDFLAGS="-L/usr/local/lib" --prefix="/usr/local"
$ make
$ sudo make install
$ src
$ sudo chmod -R u=rwX,go=rX clustalo
```
Then add the path of ClustalO into the last line the .bashrc file 
```
export PATH="/<Path to ClustalO directory on your machine>/clustal-omega-1.2.4/src:$PATH"
```

***

2\. Methodology:
----------------
![Alt text](https://github.com/Proteinea/vesba/blob/cb2606b1ac5c710be0b86b6a263255e67db4d4dd/vespa_pipeline.jpg)
*There is a main script and two auxillary scripts:*

### **1. vespa_eval *(Main Script)***

#### **1.1. Inputs:**

* It takes the **FULL path** of  wildtype sequence/s and generated sequences as a two separate **FASTA files** in addition to the path of the **cache directory** to download prott5 weights in this directory or reuse the existing weights if present in the folder. Also it takes the **number of threads** to run the functions for each sequence in parallel on different cores.

#### **1.2. Functions:**

* **fasta2dict:** Converts fasta files into Python dictionaries
* **run_vespa:** Runs *VESPA-light* on the Wildtype sequences, there are two main outputs here: 
1. CSV file for each **natural sequence** in the wildtype dataset containing SAV effect scores for the 20 possible amino acid mutations in each postion in this sequence ("/vespa_run_directory/output/")
2. FASTA file conspred_class.fasta containing the 9 class conservation score (0 highly variable, 8 Highly conserved) for each amino acid in each Wildtype sequence ("/vespa_run_directory/conspred_class.fasta")
```
>n_0
8,7,4,8,4,7,5,7,6,3,4,5,4,5,7,4,5,6,5,7,5,8,4
```
* **run_mmseq:** Runs pairwise alignment for each wildtype sequence and its corresponding generated sequence using *MMSeq2* in order to specify the positions of substitutional mutations. The output is a TAB delimited alignment file.
* **generate_msa:** Runs MSA alignment for the generated with the natural sequences using ClustalO. The output file is MSA alignment of all generated and natural sequences in FASTA format.
* **pair_map:** This function uses **vespa_pair.py** script to map SAV effect scores generated by VESPAl to the **Pairwise-aligned** generated sequence indices. This function also runs **vespa_pair.py** for each sequence in parallel based on the specified number of threads. The output is a fasta file and a csv file per each generated sequence stored in a directory called **vespa_scores**.
1. *Fasta File*: contains the pairwise alignment of each natural sequence and it corresponding generated sequence.
2. *CSV File*: contains 3 columns: **Mutation** (original amino acid type, position in natural seq, the new amino acid type in the generated seq), **Score** (SAV effect score for this mutation as predicted by VESPAl), and **Gen Seq Index** (the position of this mutation in the generated sequence)
* **msa_map:** This function uses **vespa_msa.py** script to map SAV effect scores and Conservations Classes to the indices of the **MSA-aligned** generated and natural sequences respectively. This function also runs **vespa_msa.py** for each sequence in parallel based on the specified number of threads. There are 3 main outputs:
1. *Fasta File*: Contains the MSA alignment of all the generated and natural sequences 
2. *cons_scores CSV File*: The columns are the aligned positions of natural sequences and each row is the conservation class of each position in each natural sequence.
3. *sav_scores CSV File*: The columns are the aligned positions of generated sequences and each row is the SAV effect score of each position in each generated sequence.

#### **1.3. Plots:**
* **1.3.1. Positional Conservation Class and SAV Effect Score Plot:**
The x-axis is the position in the MSA aligned natural and generated sequences, y-axis is the **Average** conservation class for natural sequences per position and the **Average** SAV effect score for generated sequences per postion. 
![Alt text](https://github.com/Proteinea/vesba/blob/70d8a7c98ffd4a5a08334a01ac337babec078957/plots/image.png)

* **1.3.2. Conservation Class vs SAV Effect Relation:** The y-axis is the Average Conservation Class of each postion in wildtype sequences and the x-axis is the Average SAV Effect Score of each postion in generated sequences. So for each MSA postion the y value represents the Average conservation class of this postion in wildtype seqs and the x value represents the Average SAV effect of mutations introduced in this postion in the generated sequences. A linear regression equation shows the relation between the two variables the relation can be negative or postive according to the slope. Also Spearman correlation is calculated for both variables. Good generated data should have inverse correlation.
![Alt text](https://github.com/Proteinea/vesba/blob/7cbd22d23044859724260366a20faa981849b242/plots/image-1.png)  

### **2. vespa_pair *(Auxillary Script)***
This script is called by vespa_eval iteratively for each wildtype-generated pair. It maps the VESPAl generated SAV effect scores for the pairwise aligned generated sequences against its corresponding natural sequence
### **2. vespa_msa *(Auxillary Script)***
This script is called by vespa_eval iteratively for each generated and natural sequence pair. It maps the VESPAl generated SAV effect scores for the MSA aligned generated sequence and Conservation Class for the corresponding Natural sequence.

* ***
3\. Usage:
----------

*   To run this code:

```plain
$ conda activate vespa
$ python vespa_eval.py -r <FULL Path of reference/natural sequence FASTA file> -g <FULL Path of generated sequence FASTA file> --cache <Path of Cache folder> --threads <Number of threads>
```
